[Feature] Support Efficient Sparse HiP Attention (InfiniteHiP) with Long-Context Generalization and KV Offloading Capabilties

daniel-geon-park opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Dear SGLang maintainers,

We implement [InfiniteHiP: Extending Language Model Context Up to 3 Million Tokens on a Single GPU](https://openreview.net/forum?id=ThpNGDBWGe), the latest version of HiP Attention from our latest ICLR 2025 paper, ["A Training-Free Sub-quadratic Cost Transformer Model Serving Framework with Hierarchically Pruned Attention"](https://openreview.net/forum?id=PTcMzQgKmn).

Just by enabling our mechanism by the flag `--enable-hip-attention`, the model is able to generalize beyond the context length it was trained on, and perform efficient sparse attention, without any noticeable degradation on the pretrained model's capabilities.

Furthermore, by using the switch `--enable-hip-kv-cache-offload`, our (experimental) KV offloading mechanism is available, which alleviates the memory issue with extremely long contexts, which enables serving up to 3 million tokens on a single L40S GPU.

The patches added to the SGLang codebase is minimal, as most of the implementation is in [our library](https://github.com/DeepAuto-AI/hip-attention/).

## Modifications

* Add a switch for HiP Attention
* Add a custom backend (hip_radix_attention.py) for HiP Attention
* Add a custom memory pool for KV offloading
* Adjust the model code (llama, qwen, exaone for now) for RoPE adjustment for HiP Attention

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

zhaochenyang20 commented:
@daniel-geon-park Sorry for waiting too long. Could rebase this? We've remind the reviewer and will review and merge it soon.

daniel-geon-park committed:
Support HiP Attention

Co-authored-by: Heejun Lee <gmlwns5176@gmail.com>
Co-authored-by: Bumsik Kim <k.bumsik@gmail.com>

# Conflicts:
#	python/sglang/srt/model_executor/cuda_graph_runner.py

daniel-geon-park commented:
> @daniel-geon-park Sorry for waiting too long. Could rebase this? We've remind the reviewer and will review and merge it soon.

Thanks! I have squashed and rebased the commits.

zhaochenyang20 commented:
Thanks! I will ask hai for help.

zhaochenyang20 commented:
@daniel-geon-park hello, could you rebase with the main branch? I asked hai for help again.

daniel-geon-park commented:
> @daniel-geon-park hello, could you rebase with the main branch? I asked hai for help again.

I rebased with the main branch again.

HaiShaw commented:
@daniel-geon-park sorry for the delay, I am doing this top of anything else.
Thanks for the patience!

HaiShaw commented:
@daniel-geon-park for the hip-attention library and kernels, do they work with both Nvidia and AMD GPUs?

daniel-geon-park commented:
> @daniel-geon-park for the hip-attention library and kernels, do they work with both Nvidia and AMD GPUs?

It only works for Nvidia GPUs at the moment.

HaiShaw commented:
@daniel-geon-park do you have this feature compatible with chunk-prefill?

daniel-geon-park commented:
> @daniel-geon-park do you have this feature compatible with chunk-prefill?

Yes indeed it is!

HaiShaw commented on python/sglang/srt/model_executor/model_runner.py:
```diff
@@ -772,7 +826,14 @@ def init_cublas(self):
     def init_attention_backend(self):
         """Init attention kernel backend."""
-        if self.server_args.attention_backend == "flashinfer":
+
```
make newer ones to the end vs. start

HaiShaw commented on python/sglang/srt/server_args.py:
```diff
@@ -114,6 +119,15 @@ class ServerArgs:
+    enable_hip_offload: bool = False
+    hip_max_mask_cache_token_size: int = 64 * 1024
+    hip_max_sa_cache_token_size: int = 8 * 1024
+
```
Can we add some comments to these values?

HaiShaw commented on python/sglang/srt/server_args.py:
```diff
@@ -688,6 +702,44 @@ def add_cli_args(parser: argparse.ArgumentParser):
+        parser.add_argument(
+            "--hip-max-mask-cache-token-size",
+            type=int,
+            default=128 * 1024,
```
This is GPU model highly dependent, do we have a way to decide it better?

daniel-geon-park replied:
I will update the interface to be less hardware-dependent by changing these parameters into `--hip-max-mask-cache-factor` and `--hip-max-sa-cache-factor`.

HaiShaw commented on python/sglang/srt/server_args.py:
```diff
@@ -688,6 +702,44 @@ def add_cli_args(parser: argparse.ArgumentParser):
+            type=int,
+            default=16 * 1024,
+            help=(
+                "On-gpu cache size of sparse attention kernels. "
```
same, better tuning mechanism?

HaiShaw reviewed (changes requested):
Thanks for the contribution!
General comments:
- Too much interface and class level changes, should look forward to some code refactor to minimize that, and avoid performance impact.
- Need to explore the options to have it on AMD/ROCm, or give log/assert info from args options on ROCm.
- Just the 3 models involved?
- What if run with DeepSeekV3/R1, etc. error out or not support? Need to handle these sorts.
- Provide example runs with commands and results, please?
- Where is HiP package to be imported, should update `pyproject`?

HaiShaw commented:
@merrymercy May you also have a look?

daniel-geon-park commented on python/sglang/srt/hf_transformers_utils.py:
```diff
@@ -117,6 +117,14 @@ def get_context_length(config):
+        if hasattr(text_config, key):
+            setattr(text_config, key, new_context_length)
+
+
```
The `update_context_length` function changes the context length in the model config so that the cached positional embeddings are long enough to handle the extended inputs.

On the other hand, rope scaling keeps the original context length in a separate variable called config.rope_scaling.original_max_position_embeddings, which is not modified here. So this should not interfere with the rope scaling.

daniel-geon-park commented on python/sglang/srt/configs/model_config.py:
```diff
@@ -80,22 +81,22 @@ def __init__(
             if context_length > derived_context_len:
-                if get_bool_env_var(
+                if is_context_extended:
+                    pass
```
It would be somewhat confusing if users get a warning about "incorrect model outputs or CUDA errors" when HiP context extension is enabled, and the environment variable is not very user-friendly. I would prefer to replace `pass` with a informative log.

daniel-geon-park commented on python/sglang/srt/configs/model_config.py:
```diff
@@ -44,6 +44,7 @@ def __init__(
         is_embedding: Optional[bool] = None,
         dtype: str = "auto",
         quantization: Optional[str] = None,
+        is_context_extended: Optional[bool] = None,
```
This is necessary to suppress the warning in line 89.

daniel-geon-park commented on python/sglang/srt/model_executor/cuda_graph_runner.py:
```diff
@@ -276,16 +279,20 @@ def can_run(self, forward_batch: ForwardBatch):
-                if self.disable_padding
-                else forward_batch.batch_size <= self.max_bs
-            )
+            if self.disable_padding:
```
I initially did this because the index type of `self.graphs` changed into a tuple, but I see there's a better way to write this. I'll rewrite this portion.

daniel-geon-park replied:
Fixed!

kbumsik commented:
> * Need to explore the options to have it on AMD/ROCm, or give log/assert info from args options on ROCm.

Hi. I am a infra engineer supporting this PR. Currently we do not have an AMD hardware to test on cloud, but I will take a look if we can get some cloud AMD hardware within the budget.

But the only cloud hardware I can find is MI300X on Runpod. Would testing on MI300X be sufficient? I would appreciate if you know any other AMD GPU cloud provider.

HaiShaw commented:
> > * Need to explore the options to have it on AMD/ROCm, or give log/assert info from args options on ROCm.
> 
> Hi. I am a infra engineer supporting this PR. Currently we do not have an AMD hardware to test on cloud, but I will take a look if we can get some cloud AMD hardware within the budget, and if @daniel-geon-park and @gmlwns2000 can work on AMD hardware.
> 
> But the only cloud hardware I can find is MI300X on Runpod. Would testing on MI300X be sufficient? I would appreciate if you know any other AMD GPU cloud provider.

@kbumsik Thanks for the question. MI300X should be sufficient, I will try to help arranging an access if you really can't find. Ping me in Slack if needed.

daniel-geon-park committed:
Improve for review comments

# Conflicts:
#	python/sglang/srt/model_executor/model_runner.py

daniel-geon-park committed:
remove rope reuse code

# Conflicts:
#	python/sglang/srt/models/llama.py
#	python/sglang/srt/models/qwen2.py

daniel-geon-park commented on python/sglang/srt/layers/radix_attention.py:
```diff
@@ -34,6 +36,8 @@ def __init__(
         sliding_window_size: int = -1,
         is_cross_attention: bool = False,
+        orig_context_len: Optional[int] = None,
+        rope: Optional[RotaryEmbedding] = None,
```
the context extension mechanism of HiP Attention requires that the RoPE is not pre-applied to the queries and keys, but instead it is applied inside the attention kernel. Therefore, we need to somehow pass the cached RoPE to the attention backend. If you have a better idea for this, then please let me know.

daniel-geon-park committed:
update exaone

daniel-geon-park committed:
fix if statement

daniel-geon-park commented:
>  * Too much interface and class level changes, should look forward to some code refactor to minimize that, and avoid performance impact.

We tried to minimize the interface level changes, but I believe this is the minimal amount of change to make our method work in SGLang. If you have some suggestions, I'd be happy to reflect those.

>  * Just the 3 models involved?
>  * What if run with DeepSeekV3/R1, etc. error out or not support? Need to handle these sorts.

That's a fair point. We'll incorporate more popular models such as DeepSeekV3/R1, and add logic to error out with unsupported models.

>  * Provide example runs with commands and results, please?

Sure. Please refer to the following page for example commands:

https://github.com/DeepAuto-AI/hip-attention/blob/deepauto/dev/docs/USAGE.sglang.md

>  * Where is HiP package to be imported, should update `pyproject`?

We are concurrently working to add the HiP package to the PyPI index, but it has not been published yet. We will update as soon as it is online.

kbumsik commented:
> @kbumsik Thanks for the question. MI300X should be sufficient, I will try to help arranging an access if you really can't find. Ping me in Slack if needed.

Thank you for answering! Unfortunately we are unable to get budget AMD GPU at the moment:

- [Runpod](https://www.runpod.io/): Says it is currently out of stock.
- [Vultr](https://www.vultr.com/products/cloud-gpu/): Currently requires long-term contract.
- [Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/gpu-accelerated/ndmi300xv5-series?tabs=sizebasic): Seems not available for new customer (we have been using AWS)

So could I possibly ask you for arranging an access? I am not sure your nickname on Slack to send a DM, so I am posting here. You can find me on SGLang slack (kbumsik@gmail.com) btw.

kbumsik commented:
> Unfortunately we are unable to get budget AMD GPU at the moment

Oh nevermind. We got one MI300X on RunPod 🚀

gmlwns2000 commented:
Hi @HaiShaw, thanks to @kbumsik, I can verify that our InfiniteHiP is working on AMD MI300X. I wrote a guide to install and run InfiniteHiP in AMD platform in this PR (https://github.com/DeepAuto-AI/hip-attention/pull/45). Thanks to great work of AMD, even CUDA UVM is compatible with AMD HIP using `hip-python` library. 

However, during testing, I found a bug (perhaps) in the Cuda graph batch size list initializer when using AMD HIP. Can you check: https://github.com/sgl-project/sglang/pull/4638 ? Extremely large batch size seems to return a memory allocation error.

Again, thank you for reviewing our PR.

daniel-geon-park committed:
add error messages for unsupported models for hip attention

daniel-geon-park committed:
Implement abstract methods

daniel-geon-park committed:
fix eagle draft cuda graph runner

daniel-geon-park committed:
support deepseek-v2

# Conflicts:
#	python/sglang/srt/models/deepseek_v2.py

zhaochenyang20 commented:
@daniel-geon-park what's the process. what blocks us to merge this?

Swipe4057 commented:
Strange error in CI:
huggingface_hub.errors.HfHubHTTPError: 429 Client Error: Too Many Requests for url: https://huggingface.co/api/models/meta-llama/Llama-3.1-8B-Instruct

[2025-03-27 05:13:48] Received sigquit from a child process. It usually means the child failed.

zhaochenyang20 commented:
@Swipe4057 I will rerun

HaiShaw commented:
> Hi @HaiShaw, thanks to @kbumsik, I can verify that our InfiniteHiP is working on AMD MI300X. I wrote a guide to install and run InfiniteHiP in AMD platform in this PR ([DeepAuto-AI/hip-attention#45](https://github.com/DeepAuto-AI/hip-attention/pull/45)). Thanks to great work of AMD, even CUDA UVM is compatible with AMD HIP using `hip-python` library.
> 
> However, during testing, I found a bug (perhaps) in the Cuda graph batch size list initializer when using AMD HIP. Can you check: #4638 ? Extremely large batch size seems to return a memory allocation error.
> 
> Again, thank you for reviewing our PR.

@gmlwns2000 PR #4638 is pre-approved, pls update and have a clean CI, then we will have it merged. Thanks.

HaiShaw commented:
@daniel-geon-park How soon we may have PyPI available? May we have `python/pyproject.toml` updated accordingly?

gmlwns2000 commented:
@HaiShaw

By "have a clean CI," do you mean that all GitHub Checks have passed? I've updated my branch, but I can't trigger the checks on the PR—it says: This workflow is awaiting approval from a maintainer in https://github.com/sgl-project/sglang/pull/4638.

Could you run the checks on that PR?

Thanks!

daniel-geon-park committed:
giupdate pyproject.toml to include hip-attn

# Conflicts:
#	python/pyproject.toml

daniel-geon-park committed:
fix test flag

daniel-geon-park committed:
move for loop over batch dim into hip-attention library

kbumsik committed:
feat: bump hip-attn

daniel-geon-park committed:
bump hip-attn version

daniel-geon-park committed:
cleanup rebase

gmlwns2000 committed:
fix

# Conflicts:
#	python/sglang/srt/models/llama4.py

daniel-geon-park committed:
fix for MLA

daniel-geon-park committed:
fix args

daniel-geon-park committed:
update tests

gmlwns2000 committed:
fix bug

gmlwns2000 committed:
pull from updated fa3 backend

daniel-geon-park committed:
bump hip-attn version

daniel-geon-park committed:
cleanup after rebase

gmlwns2000 committed:
handling chunk attention in hip

gmlwns2000 committed:
PASSKEY, you should remove before PR

daniel-geon-park committed:
remove redundant code

gmlwns2000 committed:
fix

gmlwns2000 committed:
fix

gmlwns2000 committed:
fix

# Conflicts:
#	python/sglang/srt/server_args.py

gmlwns2000 committed:
fix

Swipe4057 commented:
@daniel-geon-park Hi! I wanted to test your function on H100, but after pulling the latest changes from the master branch, I encountered two conflicts in the files python/sglang/srt/models/deepseekv2.py and python/sglang/srt/serverargs.py: needs merge. I tried to fix them myself and run it, but when launching with the Qwen2.5 model, I'm getting CUDA errors during graph capture, with a possible cause being illegal memory access in the dualstagequadratichipattention function.

gmlwns2000 committed:
cleaning

gmlwns2000 committed:
fmt

daniel-geon-park commented:
@Swipe4057 Hi, sorry for the late reply. It seems like you're running out of memory. Try adjusting --cuda-graph-max-bs and --mem-fraction-static. For example, the following runs well on 2x RTX 4090:

```python3 -m sglang.launch_server --model-path meta-llama/Llama-3.1-8B-Instruct --enable-hip-attention --context-length 210000 --cuda-graph-max-bs 16 --mem-fraction-static 0.7 --tp-size 2```

gmlwns2000 commented:
Hi, @Swipe4057 

>  I'm getting CUDA errors during graph capture, with a possible cause being illegal memory access in the dual_stage_quadratic_hip_attention function

Can you let me know which command you used to launch the sglang server? I already tested Qwen2.5 several times, and I could not reproduce the same error. (I am working on another branch `feat/llama4-sglang-pr` for sglang and `feat/llama4` for hip-attention for handling llama 4 model architecture. However, those branches are forked from this branch. I am just letting you know in case there is a potential bug in the current sglang-pr branch.)

Thanks!

gmlwns2000 committed:
support qwen2.5 vL

daniel-geon-park committed:
cleanup

daniel-geon-park commented:
@zhaochenyang20 @HaiShaw 

Bumping this PR to move it forward. Can it be merged? If not, what blocks it from being merged?

sleepwalker2017 commented:
Nice work!  How does it work for multi modal models?

gmlwns2000 commented:
Hi @sleepwalker2017 ,

We tested on Qwen2.5 VL and Llama 4 models with HiP Attention. We replace the attention mechanisms of LLM with HiP.

I am not sure this PR contains patches for handling vision models, but my working branch (`feat/delta` on `DeepAuto-AI/sglang` and `DeepAuto-AI/hip-attention`) definitely handles the vision models.

Here are our draft performance results:

| LongVideoBench (first 748 samples) | T | FA | Ours|
|--|--|--|--|
|Qwen2.5 VL 32B | 256 | 52.27 | 51.07|
|Llama4 Scout 109B | 128 | 56.15 | 54.28|

I am not sure the above result is sufficiently optimized, because I simply copied almost the same hyperparameters from language models with HiP, so probably the performance will be increased when we optimize the hyperparameters. 

For reproduce, this is command to run SGlang with vision models:

```sh
# Llama4 + HiP

SGLANG_IMAGE_PROCESSOR_FAST_DEVICE=cpu \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
HIP_DISABLE_FLASHDECODE=1 \
HIP_HEAD_REDUCE=0 \
HIP_DEBUG_LAST_DENSE=128 \
HIP_DEBUG_FORCE_DENSE_DECODE=1 \
CUDA_LAUNCH_BLOCKING=0 \
HIP_DEBUG=0 \
python -m sglang.launch_server \
  --model-path meta-llama/Llama-4-Scout-17B-16E-Instruct \
  --port 20000 --tp 8 --max-total-tokens 200000 --context-length 200000 \
  --cuda-graph-bs 1 --max-running-req 1 --chunked-prefill-size -1 \
  --hip-attention-config '{"using_extend": false, "dense_layers": [0, 1, 2, 3]}' \
  --disable-radix-cache --attention-backend flashinfer --disable-cuda-graph \
  --enable-multimodal --chat-template llama-4 --enable-hip-attention

# QwenVL 2.5 + HiP

VIDEO_MAX_PIXELS=$((128000 * 28 * 25)) \
SGLANG_IMAGE_PROCESSOR_FAST_DEVICE=cpu \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \HIP_DISABLE_FLASHDECODE=1 HIP_HEAD_REDUCE=1 HIP_DEBUG_LAST_DENSE=128 HIP_DEBUG_FORCE_DENSE_DECODE=1 CUDA_LAUNCH_BLOCKING=0 HIP_DEBUG=0 python -m sglang.launch_server --model-path Qwen/Qwen2.5-VL-72B-Instruct-AWQ --port 30000 --tp 8 --max-total-tokens 256000 --context-length 256000 --cuda-graph-bs 1 --max-running-req 1 --chunked-prefill-size -1 --hip-attention-config '{"using_extend": true, "dense_layers": [0, 1, 2, 3, 11, 19, 27, 35, 43, 51, 59, 67, 75, 79], "layers": [{"sliding_window_size": "122880", "second_stage_k": 4096, "sa_extend_backend": "streaming", "scan_extend_backend": "streaming", "stages": [{"stage_block_size_q": 64, "stage_block_stride_q": 4, "stage_chunk_size": 128, "stage_k": null, "stage_stride": 1}, {"stage_block_size_q": 64, "stage_block_stride_q": 4, "stage_chunk_size": 32, "stage_k": 32768, "stage_stride": 1}, {"stage_block_size_q": 64, "stage_block_stride_q": 1, "stage_chunk_size": 8, "stage_k": 8192, "stage_stride": 1}]}, {"sliding_window_size": 4096, "second_stage_k": 2048, "sa_extend_backend": "streaming", "scan_extend_backend": "streaming", "stages": [{"stage_block_size_q": 64, "stage_block_stride_q": 4, "stage_chunk_size": 128, "stage_k": null, "stage_stride": 1}, {"stage_block_size_q": 64, "stage_block_stride_q": 4, "stage_chunk_size": 32, "stage_k": 32768, "stage_stride": 1}, {"stage_block_size_q": 64, "stage_block_stride_q": 1, "stage_chunk_size": 8, "stage_k": 8192, "stage_stride": 1}]}]}' --disable-radix-cache --attention-backend flashinfer --disable-cuda-graph --enable-multimodal --chat-template qwen2-vl --kv-cache-dtype fp8_e5m2 --enable-hip-attention
```

gmlwns2000 commented:
@sleepwalker2017 

The table shows the model performance on [LongVideoBench](https://longvideobench.github.io/) (accuracy). We did not measure the latency on that benchmark yet, but our [paper](https://arxiv.org/abs/2502.08910) might be helpful to understand the latency of our approach.

Thanks!
